File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/auto/Weather.scala
package com.xebia.functional.xef.scala.conversation

import com.xebia.functional.xef.scala.conversation.*
import com.xebia.functional.xef.conversation.llm.openai.OpenAI
import com.xebia.functional.xef.reasoning.serpapi.Search
import com.xebia.functional.xef.prompt.Prompt

private def getQuestionAnswer(question: Prompt)(using conversation: ScalaConversation): String =
  val search: Search = Search(OpenAI.FromEnvironment.DEFAULT_CHAT, conversation, 3)
  addContext(search.search("Weather in Cádiz, Spain").get())
  promptMessage(question)

@main def runWeather: Unit =
  val question = Prompt("Knowing this forecast, what clothes do you recommend I should wear if I live in Cádiz?")
  println(conversation(getQuestionAnswer(question)).mkString("\n"))

File: examples/kotlin/src/main/kotlin/com/xebia/functional/xef/auto/Weather.kt
package com.xebia.functional.xef.auto

import com.xebia.functional.xef.auto.llm.openai.OpenAI
import com.xebia.functional.xef.auto.llm.openai.promptMessage
import com.xebia.functional.xef.prompt.Prompt
import com.xebia.functional.xef.reasoning.serpapi.Search
import io.github.oshai.kotlinlogging.KotlinLogging

suspend fun main() {
  val logger = KotlinLogging.logger("Weather")

  val question = Prompt("Knowing this forecast, what clothes do you recommend I should wear?")
  val answer = getQuestionAnswer(question)

  logger.info { answer }
}

private suspend fun getQuestionAnswer(question: Prompt): String =
  OpenAI.conversation {
    val search = Search(OpenAI.FromEnvironment.DEFAULT_CHAT, this)
    addContext(search("Weather in Cádiz, Spain"))
    promptMessage(question)
  }
